### PR TITLE
[AB#42315] Added the customizable Channel Service

### PR DESCRIPTION
**Pull request by Zehelein**
# Pull Request

## Prerequisites

- [X] The PR is targeting the right branch (`dev` for features and `master` for
      releases)
- [X] potential **release notes** to the PR description added
- [X] potential **testing notes** to the PR description added
- [ ] appropriate labels for the PR applied

## Description and release notes

The Mosaic Channel Service allows the management of channels and playlists to stream linear videos. It allows for scheduling media entities (movies and episodes) with their videos into playlists and inserting ad breaks and scheduled videos. Channels can be used to provide VOD-to-Live/FAST (Free ad-based streaming TV) playback via streaming providers like Unified Streaming.

Before this PR, the Channel Service was available as a managed Mosaic Service. We decided to open-source the Channel Service as we already did for the Vod-to-Live service to allow complete customizability of all aspects of the FAST solution. Channel and playlist validation was therefore moved directly into the Channel Service. In addition, the playlists are now published to the Catalog Service, and localization was added for the Channel and the playlist program items.

## Testing Notes

The Channel Service is the management side of the data that goes into the Catalog Service and into the Vod-to-Live Service. 

Please verify that all the test cases work if the Channel Service has localization enabled and disabled.

Please check, that you can manage channels with and without DRM protection and an assigned video so that it can be published and the publishing validation is conducted. The data of the channel should appear in the catalog service and the VOD-to-Live service should process it, and start the streaming, you can watch the stream (showing the looped placeholder video), and the streaming URLs and the DRM key ID (if DRM was enabled) is visible in the Channel Service UI in the detail station of the channel.

Create a playlist with multiple movies and episodes in it. Define ad placeholders, add videos and verify that this is visible on the Unified Streaming side when watching the playlist.

### Validation

The Channel Service validates the channel, playlist, and the used videos to
determine the eligibility of a Channel or Playlist for publishing from the
Channel Service.

#### Channel validation rules

- A placeholder video must be set for the channel. This is used when no playlist is currently active.
- A image should be set and should give a warning if this is not the case
- When localization is enabled the title must be localized and the description should be localized for all active locales

#### Playlist validation rules

- The playlist must contain at least one program.
- A warning is reported if the playlist duration exceeds 24 hours.
- The playlist duration must not exceed 25 hours.
- The playlist start date must not be older than 24 hours.
- The playlist scheduling must not start and end with an ad pod.
- All videos associated with the playlist must have at least one common stream
  format (matching resolution, frame rate, and bitrate).
- When localization is enabled all the program item titles must be localized for all active locales

#### Video validation rules

Every video assigned to a channel or playlist must pass the following validation
rules:

- Videos must use be encoded as `H264`.
- Videos must have a separate audio stream.
- Videos must be packaged in the `CMAF` format.
- If DRM protection is enabled, DRM-protected videos must have their DRM key
  identifiers defined.
- If DRM protection is disabled, videos cannot be DRM-protected.

### CODE REVIEW

File: services/entitlement/service/src/index.ts
Comment (AxTrusov) [inline services/entitlement/service/src/index.ts:120]: I would still include `/graphiql`, either here or as a separate URL. It's
inconvenient to lose a way to Ctrl+click on URL in terminal to open the graphiql
UI right away.

File: jest.config.ts
Comment (AxTrusov) [inline jest.config.ts]: Please add a space after `//` and also run `yarn script` command from monorepo
root to do the same for other files. (this file seems to be ignored by `yarn
lint` script)

File: package.json
Comment (AxTrusov) [inline package.json]: `channel-workflows` are currently included into `build` and `clean` scripts,
should be excluded for consistency

File: package.json
Comment (AxTrusov) [inline package.json]: `media-workflows` and `media-messages` are currently included into `build:fast`
and `clean:fast` scripts and should be excluded.

Also, `build:fast` currently fails because `media-messages` seems to be running
before `channel-service` build, and cleanup part removes `media-messages`
image

File: libs/media-messages/schemas/payloads/channel/common.json
Comment (AxTrusov) [inline libs/media-messages/schemas/payloads/channel/common.json]: `biginning` -> `beginning`

File: libs/media-messages/schemas/payloads/channel/events/playlist-published-event.json
Comment (AxTrusov) [inline libs/media-messages/schemas/payloads/channel/events/playlist-published-event.json:15]: why is this named content_id and not playlist_id?

File: services/catalog/service/migrations/committed/000014-channel-localizations-and-playlist.sql
Comment (AxTrusov) [inline services/catalog/service/migrations/committed/000014-channel-localizations-and-playlist.sql:18]: Since there is no migration of existing entities to the channel_localizations
table - I assume it's fine and we don't need to mentione any migration notes in
the release notes, from Managed to Customizable, correct?

File: services/entitlement/service/src/index.ts
Comment (AxTrusov) [reply services/entitlement/service/src/index.ts:120]: Same for catalog index.ts

File: services/vod-to-live/service/src/domains/smil/generator/utils.ts
Comment (AxTrusov) [inline services/vod-to-live/service/src/domains/smil/generator/utils.ts]: I think it would be more correct to change `DetailedImage` and `DetailedVideo`
to be from `media-messages` here and in all other places, since `media-messages`
now defines them and relates them to appropriate messages

File: services/channel/service/.env.template
Comment (AxTrusov) [inline services/channel/service/.env.template]: Do we still need this flag, considering DRM protection boolean field si defined
on the Channel level?

File: services/channel/service/src/publishing/common/localization/get-validation-and-localizations.ts
Comment (AxTrusov) [inline services/channel/service/src/publishing/common/localization/get-validation-and-localizations.ts]: This case bugs me just a little bit:
- Assign movie to the playlist
- Save by going back and publish the playlist
- Observe errors like `The program "Movie Title" is not fully localized: No localization value and status was found for the field "title" with the locale "bg-BG"`
- Wait for 10-15 seconds, refresh validation messages station, and observe that
  the errors are gone.

I understand that it might take some time because of the nature of how program
items are localized, and I'm not sure if there is a better way to handle it, but
in the current state, I would prefer to give the user an indication that
`localizations might not be synchronised yet, try refreshing in a couple of seconds`?

File: services/channel/service/migrations/committed/000003-added-initial-tables.sql
Comment (AxTrusov) [inline services/channel/service/migrations/committed/000003-added-initial-tables.sql:359]: I'm not sure if `video_id` should be kept at program level since
we now have an easier access to the parent entity. (Not sure about `image_id`)

Here's an example:
- Assign some video to the movie that does not match the channel placeholder
  video, and assign said movie as a program item
- Try to publish the playlist - this error will be shown `The placeholder video and playlist video(s) must each have at least one stream with the same resolution, frame rate, and bitrate.`
- It felt logical to go to movie details, and assign a different video. But this
  would not remove the error above if you try to publish the playlist again. You
  would need to go to the playlist, unasign related movie, save changes, and
  assign it again - this would update the stored video_id, which would make the
  validation use updated `video_id`

Or maybe we need some events in the media service, e.g. image/video changed -
which channel service would pick up and update relations for its program items?

**Comment by AxTrusov** [comment]
Whenever I try to unpublish the playlist - I v2l starts throwing `The specified blob does not exist.` Is it only happening for me?

File: services/vod-to-live/service/.env.template
Comment (AxTrusov) [inline services/vod-to-live/service/.env.template:7]: Not related to this file, but general comment:

Whenever I try to unpublish the playlist - v2l starts throwing `The specified blob does not exist.`error. Is it only happening for me?

Might be related to PlaylistUnpublished having an inconsistent payload id values:
```
      payload: {
        channel_id: 'f880d4e3-7024-49c5-b80c-693e042259c2',
        content_id: 'playlist-ddafc1b3-ae8c-47fc-a18d-093d0edc7df6'
      },
```

Comment (AxTrusov) [reply services/vod-to-live/service/.env.template:7]: The stream also continues to play, so probably something should be fixed here. (unpublishing the channel works and stream "stops")

File: services/channel/service/migrations/current.sql
Comment (AxTrusov) [inline services/channel/service/migrations/current.sql]: These contents are duplicating the initial tables migration and should be dropped

File: services/channel/workflows/src/stations/Channels/ChannelVideoManagement/ChannelVideoManagement.tsx
Comment (AxTrusov) [inline services/channel/workflows/src/stations/Channels/ChannelVideoManagement/ChannelVideoManagement.tsx]: Are you sure we want to apply the `Approved` state to filters? I think only movie publishing validation currently cares about it, so if you assigned an un-approved video as channel placeholder - it will work just fine.

File: services/channel/service/src/index.ts
Comment (AxTrusov) [inline services/channel/service/src/index.ts:75]: I think somewhere around here, to align with `media-service` implementation, we should have an `await updateConfigWithActualLocalizationAvailability(config, logger);`

File: services/channel/service/src/index.ts
Comment (AxTrusov) [inline services/channel/service/src/index.ts]: `registerTypes` already called above, one of the calls should be dropped

File: services/channel/service/package.json
Comment (AxTrusov) [inline services/channel/service/package.json]: Please align the version with other services.
image

File: services/channel/service/package.json
Comment (AxTrusov) [inline services/channel/service/package.json]: "async-mutex" can be dropped as it's not used anywhere

File: services/channel/service/package.json
Comment (AxTrusov) [reply services/channel/service/package.json]: Same for `axios` and potentially other libs.
image

File: services/channel/service/package.json
Comment (AxTrusov) [reply services/channel/service/package.json]: "generate-password" not used as well. Would check other libs too
image

File: services/channel/service/src/common/setup/register-cue-points.ts
Comment (AxTrusov) [inline services/channel/service/src/common/setup/register-cue-points.ts]: I would rename the filename to `register-video-cue-point-types.ts` to align with the function naming.

File: services/channel/service/src/common/setup/register-localization-definitions.ts
Comment (AxTrusov) [inline services/channel/service/src/common/setup/register-localization-definitions.ts]: I would rename the filename to `register-localization-entity-definitions.ts` to align with the function naming.

File: services/channel/service/src/messaging/transactional-outbox-inbox/ensure-transactional-slots-and-publications-exist.ts
Comment (AxTrusov) [inline services/channel/service/src/messaging/transactional-outbox-inbox/ensure-transactional-slots-and-publications-exist.ts]: I think this whole file can be dropped since we are not using logical replication, right?

File: services/channel/service/src/graphql/postgraphile-middleware.ts
Comment (AxTrusov) [inline services/channel/service/src/graphql/postgraphile-middleware.ts]: If we really need altair endpoint - please add `altair-express-middleware` as an
explicit dependency to the project. Also, I wonder if there is a tool out there
that can go through the project, check all used 3-rd party dependencies, and
check which of them are not defined in `package.json`... If there is - we might
want to introduce that to our PR build pipelines.

File: services/channel/service/src/messaging/register-messaging.ts
Comment (AxTrusov) [inline services/channel/service/src/messaging/register-messaging.ts]: Everything Workflows-related can be dropped from this file.

**Comment by AxTrusov** [comment]
In general - checked that both DRM-protected and not protected streams are playable, manageable and localizable, all seems reasonable. Will wait for comments to be resolved before starting a second review round.

**Review by AxTrusov — CHANGES_REQUESTED** [review]
In general - checked that both DRM-protected and not protected streams are playable, manageable and localizable, all seems reasonable. Will wait for comments to be resolved before starting a second review round.

File: services/catalog/service/migrations/committed/000014-channel-localizations-and-playlist.sql
Comment (Zehelein) [reply services/catalog/service/migrations/committed/000014-channel-localizations-and-playlist.sql:18]: Channels have not much data so they can easily be recreated. And playlist must anyways be recreated for every day so there is no automatic migration planned.

File: package.json
Comment (Zehelein) [reply package.json]: Changing it into two parts: the media only part (media, catalog, and entitlement) and all (both media and FAST). The FAST service requires the media service thus such a split would be best IMO.

File: services/channel/service/src/graphql/postgraphile-middleware.ts
Comment (Zehelein) [reply services/channel/service/src/graphql/postgraphile-middleware.ts]: `npx depcheck` - it helps to find missing ones and suggests ones that are not used. But it is not perfect e.g. we use the GraphQL codegene but only as part of our scripts so it is not possible for the tool to find it.

File: services/channel/service/migrations/committed/000003-added-initial-tables.sql
Comment (Zehelein) [reply services/channel/service/migrations/committed/000003-added-initial-tables.sql:359]: That could be a potential future improvement. The problem is that somebody may have already filled some cue points. If we just switch the video all the filled cue points are gone. So the workflow would be to select a different video for your movie. Then add that movie (again) to the playlist as a program item. Add the same ad-spots and placeholder videos from the current program item to the new program item. Only then delete the old program item. 

Potentially that step could be based on some "merge" option where pre and post cue points are directly migrated and the editor could then select from which old cue point to which new cue point the already added items should be used. But that is quite a bit of work.

File: services/vod-to-live/service/.env.template
Comment (Zehelein) [reply services/vod-to-live/service/.env.template:7]: good point - the channel ID lacked the "channel-" prefix in that call but was included in the initial publish call.

File: services/channel/service/migrations/committed/000003-added-initial-tables.sql
Comment (AxTrusov) [reply services/channel/service/migrations/committed/000003-added-initial-tables.sql:359]: Future improvement is fine with me,

File: services/channel/service/src/publishing/common/localization/get-validation-and-localizations.ts
Comment (AxTrusov) [inline services/channel/service/src/publishing/common/localization/get-validation-and-localizations.ts]: Minor adjustment needed here and in the test: `will copied` -> `will be copied` or `will be copied over`

File: services/channel/workflows/package.json
Comment (AxTrusov) [inline services/channel/workflows/package.json]: Ideally, I would compare the workflows versions between media workflows and channel workflows and align them.
Minimally, I would align the piral-related package versions, to eliminate a warning during build:all script run:
image

File: services/vod-to-live/service/mosaic-hosting-deployment-manifest.yaml
Comment (EmperorRXF) [inline services/vod-to-live/service/mosaic-hosting-deployment-manifest.yaml:34]: is this going to point to the custom channel-service deployment ?

if yes, we have a similar reference from ent-service -> catalog-service, and there's a placeholder value we could use for the DNS value of another custom service

File: services/vod-to-live/service/mosaic-hosting-deployment-manifest.yaml
Comment (EmperorRXF) [inline services/vod-to-live/service/mosaic-hosting-deployment-manifest.yaml:38]: think we are missing the `AZURE_BLOB_CONTAINER_NAME`?

File: services/vod-to-live/service/mosaic-hosting-deployment-manifest.yaml
Comment (EmperorRXF) [reply services/vod-to-live/service/mosaic-hosting-deployment-manifest.yaml:34]: (or maybe it refers to some USP URL - then pls ignore)

**Review by AxTrusov — APPROVED** [review]

File: scripts/quick-deploy-service.ts
Comment (EmperorRXF) [inline scripts/quick-deploy-service.ts:214]: is vod-2-live not buildable via hosting v3?

File: services/vod-to-live/service/mosaic-hosting-deployment-manifest.yaml
Comment (Zehelein) [reply services/vod-to-live/service/mosaic-hosting-deployment-manifest.yaml:34]: The second one - this is the Unified Streaming Virtual Channels setup which consists of multiple micro-services. The one here is the URL to the Unified Streaming management API.

File: scripts/quick-deploy-service.ts
Comment (Zehelein) [reply scripts/quick-deploy-service.ts:214]: Good point. So far it was not but I will add it.

**Pull request merged**
